academic-services/knowledge-exploration-service/tutorial-grammar-design.md

```diff
@@ -0,0 +1,401 @@
+
+### Support composite entity attribute queries
+
+We can improve search accuracy by creating a grammar that considers search intents. A search query containing an author name and an affiliation name may have multiple intents. An intent can be "search for papers based on an author and an institution that the author has collaborated with", or "search for papers based on an author who is affiliated with an institution", etc. In the sample grammar, we create a special rule to recognize the intent, "search for papers based on an author who is affiliated with an institution". This enables the grammar to recognize the intent for queries like "papers by iheng chen while at national sun yat sen university" and return results with more accuracy.
+
+The search grammar below handles the natural language search query format for extracting the intent.
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_composite_entity_attribute_search":::
+
+We can then modify ther return query according to the recognized intent. Instead of forming the query constraint like
+
+```cmd
+And(
+  Composite(AuthorAffiliation.AuthorName="..."),
+  Composite(AuthorAffiliation.AffiliationName = "...")
+)
+```
+
+we will use the following to specifiy stronger query constraint. See [Composite function](reference-semantic-functions.md#composite-function) query for more details.
+
+```cmd
+Composite(
+  And(AuthorAffiliation.AuthorName="..."),
+  And(AuthorAffiliation.AffiliationName = "...")
+)
+```
+
+Lastly, we want to ensure the recognized intent is a valid intent. We achieve this by calling [`HasResult`](reference-semantic-functions.md#hasresult-function) using the output query and mark all the interpretations that can't retrieve query results as invalid using the [`AssertEquals`](reference-semantic-functions.md#assertequals-function) semantic function.
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_ensure_query_has_results":::
+
+### Support keyword based queries
+
+Keyword based queries are queries that contain keywords from entities attributes. For example, a user may search for a paper using terms from the paper's title or abstract.
+
+We use a `<item repeat="1-INF">` element to process keyword based queries using abstract and title words.
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_partial_attribute_search":::
+
+Notice that we have heavier penalties (-3 and -4) associated with title and abstract keyword search compared to other grammar rules in the sample grammar (i.e. author and affiliation searches have a penalty of -2). This is designed to penalize title/abstract keyword searches if semantic attribute searches yield results.
```

Comment:
>This is designed to penalize title/abstract keyword searches if semantic attribute searches yield results [](start = 215, length = 105)

"We do this so that semantic attribute matches will be favored over keyword matches when interpreting the user query"